src/pandas_profiling/model/dataframe_wrappers.py

```diff
@@ -0,0 +1,599 @@
+         memory usage of each column in bytes.
+
+        """
+        pass
+
+    @abstractmethod
+    def head(self, n: int):
+        """
+        Get top n rows of dataframe.
+        Only call within singledispatch methods
+
+        Args:
+            n: top n rows
+
+        Returns: dataframe with only top n rows
+
+        """
+        pass
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Get the number of rows in a dataframe as an int
+
+        Returns: number of rows in column
+
+        """
+        pass
+
+    @abstractmethod
+    def __getitem__(self, key):
+        pass
+
+
+class PandasDataFrame(GenericDataFrame):
+    """
+    This class is the abstract layer for the pandas dataframe.
+
+    It implements the check_if_corresponding_engine function to check if an obj is a pandas dataframe and can be wrapped.
+
+    It implements the get_pandas_df method to enable singledispatch wrapped functions to
+    operate on the internal pandas dataframe without breaking abstraction
+
+    """
+
+    engine = "pandas"
+
+    def __init__(self, df):
+        super().__init__()
+        # self.df holds the underlying data object
+        self.df = df
+        self.n_rows = len(self.df)
+
+    @staticmethod
+    def check_if_corresponding_engine(obj) -> bool:
+        """
+        Check if pandas dataframe using isinstance. More pythonic way of checking as opposed to spark type check.
+
+        Args:
+            obj: check if object is a pandas dataframe
+
+        Returns: True if the __module__ and __name__ of object matches pandas dataframe, else false
+        """
+        return isinstance(obj, pd.DataFrame)
+
+    @staticmethod
+    def preprocess(df: pd.DataFrame) -> pd.DataFrame:
+        """Preprocess the dataframe
+
+        - Appends the index to the dataframe when it contains information
+        - Rename the "index" column to "df_index", if exists
+        - Convert the DataFrame's columns to str
+
+        Args:
+            df: the pandas DataFrame
+
+        Returns:
+            The preprocessed DataFrame
+        """
+
+        # Treat index as any other column
+        if (
+            not pd.Index(np.arange(0, len(df))).equals(df.index)
+            or df.index.dtype != np.int64
+        ):
+            df = df.reset_index()
+
+        # Rename reserved column names
+        df = rename_index(df)
+
+        # Ensure that columns are strings
+        df.columns = df.columns.astype("str")
+        return df
+
+    @property
+    def columns(self) -> List[str]:
+        return self.df.columns
+
+    @property
+    def empty(self) -> bool:
+        return self.df.empty
+
+    def get_duplicate_rows_count(self, subset) -> int:
+        """
+
+        Args:
+            subset:
+            keep:
+
+        Returns:
+
+        """
+        return len(self.df[self.df.duplicated(subset=subset, keep="first")])
```

Comment:
You can avoid the index & copy here

`return self.df.duplicate(subset=subset).sum()`